File: runinator-command-center/src/core/realtime/event-stream-client.ts
import { buildWebSocketProtocols, buildWebSocketUrl } from "./websocket-url";
import type { EventStreamRouter, ServerEvent } from "./event-router";
import { ReconnectBackoff } from "./reconnect-backoff";

const FALLBACK_INTERVAL = 30000;
const CONNECT_TIMEOUT = 5000;
const SOCKET_OPEN = 1;

export type EventStreamState = "disconnected" | "connecting" | "connected" | "fallback";

export type EventSocket = Pick<
  WebSocket,
  "onopen" | "onmessage" | "onclose" | "onerror" | "readyState" | "close"
>;
export interface WebSocketFactory {
  create(url: string, protocols: string[]): EventSocket;
}
export interface EventStreamTimers {
  setTimeout(callback: () => void, delay: number): number;
  clearTimeout(handle: number): void;
  setInterval(callback: () => void, delay: number): number;
  clearInterval(handle: number): void;
}
const browserSockets: WebSocketFactory = {
  create: (url, protocols) => new WebSocket(url, protocols),
};
const browserTimers: EventStreamTimers = {
  setTimeout: (callback, delay) => window.setTimeout(callback, delay),
  clearTimeout: (handle) => {
    window.clearTimeout(handle);
  },
  setInterval: (callback, delay) => window.setInterval(callback, delay),
  clearInterval: (handle) => {
    window.clearInterval(handle);
  },
};

export interface EventStreamClientOptions {
  getServiceUrl: () => string | null;
  getServiceKnown: () => boolean;
  onStateChange: (state: EventStreamState) => void;
  onFallbackTick: () => void;
  router: EventStreamRouter;
  sockets?: WebSocketFactory;
  timers?: EventStreamTimers;
}

export class EventStreamClient {
  private ws: EventSocket | null = null;
  private fallbackTimer: number | null = null;
  private reconnectTimer: number | null = null;
  private connectTimer: number | null = null;
  private connectionId = 0;
  private readonly backoff = new ReconnectBackoff();

  private readonly sockets: WebSocketFactory;
  private readonly timers: EventStreamTimers;
  constructor(private readonly options: EventStreamClientOptions) {
    this.sockets = options.sockets ?? browserSockets;
    this.timers = options.timers ?? browserTimers;
  }

  connect() {
    this.clearReconnectTimer();
    this.clearConnectTimer();
    const currentConnection = ++this.connectionId;

    if (this.ws?.readyState === SOCKET_OPEN) {
      this.ws.close();
    }

    this.ws = null;
    const serviceUrl = this.options.getServiceUrl();

    if (!serviceUrl) {
      this.startFallback();
      return;
    }

    this.options.onStateChange("connecting");
    const url = buildWebSocketUrl(serviceUrl, "/ws/events");
    const socket = this.sockets.create(url, buildWebSocketProtocols());
    this.ws = socket;
    this.connectTimer = this.timers.setTimeout(() => {
      if (currentConnection !== this.connectionId) {
        return;
      }

      // `close()` during CONNECTING produces a noisy browser error. Invalidate this attempt and
      // let a late upgrade close itself in `onopen`, where closing is intentional and quiet.
      this.connectionId += 1;

      if (this.ws === socket) {
        this.ws = null;
      }

      this.startFallback();
    }, CONNECT_TIMEOUT);

    this.ws.onopen = () => {
      if (currentConnection !== this.connectionId) {
        socket.close();
        return;
      }

      this.clearConnectTimer();
      this.backoff.reset();
      this.options.onStateChange("connected");
      this.stopFallback();
    };

    this.ws.onmessage = ({ data }: MessageEvent<string>) => {
      if (currentConnection !== this.connectionId) {
        return;
      }

      try {
        this.options.router.route(JSON.parse(data) as ServerEvent);
      } catch {
        /* ignore malformed payloads */
      }
    };

    this.ws.onclose = () => {
      if (currentConnection !== this.connectionId) {
        return;
      }

      this.clearConnectTimer();
      this.ws = null;
      this.startFallback();

      if (this.options.getServiceKnown()) {
        this.reconnectTimer = this.timers.setTimeout(() => {
          this.connect();
        }, this.backoff.next());
      }
    };

    this.ws.onerror = () => {
      if (currentConnection !== this.connectionId) {
        return;
      }

      if (socket.readyState === SOCKET_OPEN) {
        this.clearConnectTimer();
        socket.close();
      }
    };
  }

  disconnect() {
    this.connectionId += 1;
    this.clearReconnectTimer();
    this.clearConnectTimer();
    this.backoff.reset();

    if (this.ws?.readyState === SOCKET_OPEN) {
      this.ws.close();
    }

    this.ws = null;
    this.stopFallback();
    this.options.onStateChange("disconnected");
  }

  private startFallback() {
    if (this.fallbackTimer !== null) {
      return;
    }

    this.options.onStateChange("fallback");
    this.fallbackTimer = this.timers.setInterval(() => {
      this.options.onFallbackTick();
    }, FALLBACK_INTERVAL);
  }

  private stopFallback() {
    if (this.fallbackTimer !== null) {
      this.timers.clearInterval(this.fallbackTimer);
      this.fallbackTimer = null;
    }
  }

  private clearReconnectTimer() {
    if (this.reconnectTimer === null) {
      return;
    }

    this.timers.clearTimeout(this.reconnectTimer);
    this.reconnectTimer = null;
  }

  private clearConnectTimer() {
    if (this.connectTimer === null) {
      return;
    }

    this.timers.clearTimeout(this.connectTimer);
    this.connectTimer = null;
  }
}

File: runinator-command-center-flutter/lib/core/realtime/event_stream_client.dart
// port of core/realtime/event-stream-client.ts.
//
// dart's web_socket_channel exposes connection success/failure through the
// `ready` future rather than an onopen callback; that future drives the
// "connecting -> connected" transition here in place of the ts source's
// WebSocket.onopen handler. everything else (fallback polling, reconnect
// delay, connect timeout race, connection-id fencing against stale
// callbacks) is ported verbatim.

import 'dart:async';
import 'dart:convert';

import 'package:web_socket_channel/web_socket_channel.dart';

import 'event_router.dart';
import 'websocket_url.dart';

const Duration _reconnectDelay = Duration(milliseconds: 3000);
const Duration _fallbackInterval = Duration(milliseconds: 30000);
const Duration _connectTimeout = Duration(milliseconds: 5000);

enum EventStreamState { disconnected, connecting, connected, fallback }

class EventStreamClientOptions {
  const EventStreamClientOptions({
    required this.getServiceUrl,
    required this.getServiceKnown,
    required this.onStateChange,
    required this.onFallbackTick,
    required this.router,
  });

  final String? Function() getServiceUrl;
  final bool Function() getServiceKnown;
  final void Function(EventStreamState state) onStateChange;
  final void Function() onFallbackTick;
  final EventStreamRouter router;
}

class EventStreamClient {
  EventStreamClient(this._options);

  final EventStreamClientOptions _options;

  WebSocketChannel? _channel;
  StreamSubscription<Object?>? _subscription;
  Timer? _fallbackTimer;
  Timer? _reconnectTimer;
  Timer? _connectTimer;
  int _connectionId = 0;

  void connect() {
    _clearReconnectTimer();
    _clearConnectTimer();
    final serviceUrl = _options.getServiceUrl();

    if (serviceUrl == null) {
      _startFallback();
      return;
    }

    final currentConnection = ++_connectionId;
    _options.onStateChange(EventStreamState.connecting);
    final url = buildWebSocketUrl(serviceUrl, '/ws/events');

    final channel = WebSocketChannel.connect(Uri.parse(url));
    _channel = channel;

    _connectTimer = Timer(_connectTimeout, () {
      if (currentConnection != _connectionId) {
        return;
      }

      _channel?.sink.close();
      _startFallback();
    });

    channel.ready.then((_) {
      if (currentConnection != _connectionId) {
        return;
      }

      _clearConnectTimer();
      _options.onStateChange(EventStreamState.connected);
      _stopFallback();
    }).catchError((Object _) {
      // connection failed to establish; the stream's onDone below drives fallback/reconnect.
    });

    _subscription = channel.stream.listen(
      (data) {
        if (currentConnection != _connectionId) {
          return;
        }

        try {
          _options.router.route(jsonDecode(data as String) as ServerEvent);
        } catch (_) {
          // ignore malformed payloads
        }
      },
      onDone: () {
        if (currentConnection != _connectionId) {
          return;
        }

        _clearConnectTimer();
        _channel = null;
        _startFallback();

        if (_options.getServiceKnown()) {
          _reconnectTimer = Timer(_reconnectDelay, connect);
        }
      },
      onError: (Object _) {
        if (currentConnection != _connectionId) {
          return;
        }

        _clearConnectTimer();
        _channel?.sink.close();
      },
      cancelOnError: true,
    );
  }

  void disconnect() {
    _connectionId += 1;
    _clearReconnectTimer();
    _clearConnectTimer();
    _subscription?.cancel();
    _channel?.sink.close();
    _channel = null;
    _stopFallback();
    _options.onStateChange(EventStreamState.disconnected);
  }

  void _startFallback() {
    if (_fallbackTimer != null) {
      return;
    }

    _options.onStateChange(EventStreamState.fallback);
    _fallbackTimer = Timer.periodic(_fallbackInterval, (_) => _options.onFallbackTick());
  }

  void _stopFallback() {
    _fallbackTimer?.cancel();
    _fallbackTimer = null;
  }

  void _clearReconnectTimer() {
    _reconnectTimer?.cancel();
    _reconnectTimer = null;
  }

  void _clearConnectTimer() {
    _connectTimer?.cancel();
    _connectTimer = null;
  }
}
